### PR TITLE
ctrl: Result, Submission

### PR DESCRIPTION
**Pull request by victorbalan**
* Add possibility to finish a task by setting the
   final submission in the result

### CODE REVIEW

File: controllers/result.go [hunk 1]
```diff
@@ -74,7 +74,9 @@ func CreateResult(ctx context.Context, w http.ResponseWriter, r *http.Request) (
 		return http.StatusInternalServerError, err
 	}
 
-	result := model.Result{Challenge: key, StartTimes: make([]time.Time, len(challenge.Tasks))}
+	result := model.Result{Challenge: key,
+		StartTimes:       make([]time.Time, len(challenge.Tasks)),
+		FinalSubmissions: make([]*datastore.Key, len(challenge.Tasks))}
```
Comment (flowlo) [inline controllers/result.go]: I think it should be either everything in one line, or every property on it's own line:

```go
result := model.Result{
	Challenge:        key,
	StartTimes:       make([]time.Time, len(challenge.Tasks)),
	FinalSubmissions: make([]*datastore.Key, len(challenge.Tasks)),
}
```

File: controllers/submission.go [hunk 2]
```diff
@@ -71,3 +71,56 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		return http.StatusUnauthorized, nil
+	}
+	if r.Method != "POST" {
+		return http.StatusMethodNotAllowed, nil
+	}
+
+	resultKey, err := datastore.DecodeKey(mux.Vars(r)["resultKey"])
+
+	if !util.HasParent(p.UserKey, resultKey) {
+		return http.StatusBadRequest, errors.New("Cannot submit answer for other users")
```
Comment (flowlo) [inline controllers/submission.go]: [Error strings should not be capitalized](https://github.com/golang/lint/blob/master/lint.go#L1163).

File: controllers/submission.go [hunk 3]
```diff
@@ -71,3 +71,56 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+	}
+
+	taskKey, err := datastore.DecodeKey(mux.Vars(r)["taskKey"])
+	var result model.Result
+	if err = datastore.Get(ctx, resultKey, &result); err != nil {
+		return http.StatusInternalServerError, err
+	}
+	var challenge model.Challenge
```
Comment (flowlo) [inline controllers/submission.go]: Could use a newline between the error checking for the previous call and initializing the next thing.

File: controllers/submission.go [hunk 4]
```diff
@@ -71,3 +71,56 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		return http.StatusInternalServerError, err
+	}
+
+	var submissionKeys []*datastore.Key
+	if submissionKeys, err = model.NewQueryForCodeSubmission().
+		Ancestor(resultKey).
+		Order("-Start").
+		KeysOnly().
+		Limit(1).
+		GetAll(ctx, nil); err != nil {
```
Comment (flowlo) [inline controllers/submission.go]: This call chain is so big, it could be moved out of the `if`.

**Comment by flowlo** [comment]
Why do we need setting submissions final in two places?

**Comment by victorbalan** [comment]
we need it for the current coding status

**Comment by flowlo** [comment]
Please elaborate.

File: controllers/submission.go [hunk 5]
```diff
@@ -71,3 +71,56 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+	if err = datastore.Get(ctx, resultKey, &result); err != nil {
+		return http.StatusInternalServerError, err
+	}
+	var challenge model.Challenge
+	if err = datastore.Get(ctx, result.Challenge, &challenge); err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	var submissionKeys []*datastore.Key
+	if submissionKeys, err = model.NewQueryForCodeSubmission().
```
Comment (flowlo) [inline controllers/submission.go]: This only works for `CodeSubmission`, but it should work with all kinds of submissions.

File: controllers/submission.go [hunk 6]
```diff
@@ -71,3 +71,56 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		Ancestor(resultKey).
+		Order("-Start").
```
Comment (flowlo) [inline controllers/submission.go]: `Start` is only present for `CodeSubmission`. Why not use `Time`, which is present in every `Submission`?

**Comment by victorbalan** [comment]
if a company wants to see live how a coder is doing we need to display somehow the `task-finished` event, so we need to persist this somehow.

File: controllers/submission.go (hunk 6 above)
Comment (victorbalan) [inline controllers/submission.go]: Time is not used anywere. I`ll add the time setting in this commit

Comment (flowlo) [inline controllers/submission.go]: > Why not use `Time` ...?

with response

> Time is not used anywere.

makes no sense to me.

Comment (victorbalan) [inline controllers/submission.go]: Time just exists there in the entity. It is not set anywhere. I`ll handle that in this commit.

**Comment by flowlo** [comment]
But the request to this route will still be made by the coder. So how does it help the company?

> the `task-finished` event

What's that, where is it defined?

**Comment by victorbalan** [comment]
It is defined here
https://github.com/coduno/app/commit/5d8cc761c94e2c9987346c26e426d6456ed8a118#diff-af6e8bb5482f4672a7168c8bbc9fe0dfR135

It will help the company knowing wich of the stundents finished faster  task in real time

**Comment by victorbalan** [comment]
I`ll rebase to fix the conflicts.

**Comment by flowlo** [comment]
OK. So it should do exactly the same thing as [this part of CreateFinalResult](https://github.com/coduno/api/blob/master/controllers/result.go#L139-L158). shouldn't it? We are duplicating code.

**Comment by victorbalan** [comment]
i did not have that when i coded

**Comment by flowlo** [comment]
> i did not have that when i coded.

It's [22 hours old](https://github.com/coduno/api/pull/38)!

**Comment by victorbalan** [comment]
And one more thing. [this part of CreateFinalResult](https://github.com/coduno/api/blob/master/controllers/result.go#L139-L158) should happen only if the finalSubmissions are not set. Because GET on /results/{key} is made to GET a result by key and a lot of companies will want to do that. That means everytime we load a result we set the finalSubmissions and also we [compute the results](https://github.com/coduno/api/blob/master/controllers/result.go#L132) every time.

**Comment by flowlo** [comment]
Please file an issue. This is out of scope for this PR.

**Pull request merged**